### PR TITLE
Add Exception Handling to File Name for Telemetry Caching

### PR DESCRIPTION
**Pull request by JacksonWeber**
https://github.com/microsoft/ApplicationInsights-node.js/issues/1230

Append the process ID to the file name created for holding disk cached telemetry. This should resolve the issue with multiple Azure Functions cores attempting to read/write/delete the same file when functions are scaled to use multiple cores.

Extended this logic outside of Azure Functions so that in any case where the SDK could be run concurrently we create distinct cache files.

### CODE REVIEW

File: Library/FileSystemHelper.ts [hunk 1]
```diff
@@ -39,38 +40,52 @@ export const confirmDirExists = async (directory: string): Promise<void> => {
...
+            if (fileStats.isFile()) {
+                totalSize += fileStats.size;
+            }
         }
+        return totalSize;
+    } catch {
+        Logging.warn(`Failed to get directory size for ${directory}`);
+        return totalSize;
     }
-    return totalSize;
```
Comment (hectorhdzg) [inline Library/FileSystemHelper.ts:52]: You can keep this one and remove the other two returns

File: Library/FileSystemHelper.ts [hunk 2]
```diff
@@ -39,38 +40,52 @@ export const confirmDirExists = async (directory: string): Promise<void> => {
...
+        let files = fs.readdirSync(directory);
+        for (let i = 0; i < files.length; i++) {
+            totalSize += fs.statSync(path.join(directory, files[i])).size;
+        }
+        return totalSize;
+    } catch {
+        Logging.warn(`Failed to get directory size synchronously for ${directory}`)
+        return totalSize;
     }
-    return totalSize;
```
Comment (hectorhdzg) [inline Library/FileSystemHelper.ts:64]: Same here

File: Library/FileSystemHelper.ts [hunk 3]
```diff
@@ -39,38 +40,52 @@ export const confirmDirExists = async (directory: string): Promise<void> => {
...
-    if (fileStats.isFile()) {
-        return fileStats.size;
+    try {
+        const fileStats = await statAsync(filePath);
+        if (fileStats.isFile()) {
+            return fileStats.size;
+        }
+    } catch {
+        Logging.warn(`Failed to get file size for ${filePath}`);
     }
```
Comment (hectorhdzg) [inline Library/FileSystemHelper.ts:89]: Should you return something here in failure case?

File: Library/Sender.ts [hunk 4]
```diff
@@ -428,7 +429,7 @@ class Sender {
         try {
              //create file - file name for now is the timestamp, a better approach would be a UUID but that
             //would require an external dependency
-            var fileName = new Date().getTime() + ".ai.json";
+            var fileName = `${new Date().getTime()}.${this._processId}.ai.json`;
```
Comment (hectorhdzg) [inline Library/Sender.ts:432]: How many digiths is the processId? I wonder if this could cause issues if the file name gets very long.

Comment (hectorhdzg) [reply Library/Sender.ts:432]: We may just drop the date now right?, maybe use a UUID like the comment below?

Comment (JacksonWeber) [reply Library/Sender.ts:432]: Looks like the max length for both Windows and Linux should be 255 characters for a file path. I'll add some tests for long file names, but assuming the above holds true. I'm comfortable using up a max of < 8% of our file name length for the PID.

Comment (JacksonWeber) [reply Library/Sender.ts:432]: > We may just drop the date now right?, maybe use a UUID like the comment below?

Are you suggesting we just drop the PID & Date approach and transfer entirely to using the UUID package?

File: Library/FileSystemHelper.ts (hunk 3 above)
Comment (JacksonWeber) [reply Library/FileSystemHelper.ts:88]: I'll return -1 if the file does not exist and update the function desc.

File: Library/Sender.ts (hunk 4 above)
Comment (JacksonWeber) [reply Library/Sender.ts:432]: Is relying on the external dependency no longer a concern?

**Review by hectorhdzg — APPROVED** [review]
LGTM

**Pull request merged**